### PR TITLE
8264940: java/lang/invoke/6998541/Test6998541.java failed "guarantee(ik->is_initialized()) failed: java/lang/Byte$ByteCache must be initialized"

### PR DESCRIPTION
**Pull request by Wanghuang-Huawei**
*  The main reason is that in this situation the `BoxCache` is not initialized before `uncommon_trap`.
*  `java.lang.Byte$ByteCache` is loaded without initializing here 
* I think that all `*Cache` classes should be loaded in this situation.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264940](https://bugs.openjdk.java.net/browse/JDK-8264940): java/lang/invoke/6998541/Test6998541.java failed "guarantee(ik->is_initialized()) failed: java/lang/Byte$ByteCache must be initialized"

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 3cde7f945997dffc3eeff283b0d1b2c352869de5
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Contributors
 * Wu Yan `<wuyan34@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3411/head:pull/3411` \
`$ git checkout pull/3411`

Update a local copy of the PR: \
`$ git checkout pull/3411` \
`$ git pull https://git.openjdk.java.net/jdk pull/3411/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3411`

View PR using the GUI difftool: \
`$ git pr show -t 3411`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3411.diff">https://git.openjdk.java.net/jdk/pull/3411.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 1]
```diff
@@ -976,7 +976,7 @@ class BooleanBoxCache : public BoxCacheBase<java_lang_Boolean> {
 protected:
   static BooleanBoxCache *_singleton;
   BooleanBoxCache(Thread *thread) {
-    InstanceKlass* ik = find_cache_klass(java_lang_Boolean::symbol());
+    InstanceKlass* ik = find_cache_klass(java_lang_Boolean::symbol(), thread);
```
Comment (coleenp) [inline src/hotspot/share/runtime/deoptimization.cpp:979]: If this throws a Java exception, you have to handle it here, you can't ignore it with thread.  You should initialize these classes during initialization.  You shouldn't be doing this in deoptimization.

**Review by coleenp — CHANGES_REQUESTED** [review]

**Comment by dholmes-ora** [comment]
To evaluate what is the right fix for this problem I need to understand how  JDK-8261137 caused the problem. Can you explain that please? When was this class normally initialized?

Simply initializing the class at the point where you thought it already should have been initialized may not be the right solution, and as Coleen has pointed out you have to deal with any exceptions that may occur during that initialization.

Thanks,
David

**Comment by vnkozlov** [comment]
The issue happens because the test was run with `-Xcomp` flag. The method was not executed in Interpreter and class was not initialized before the method is JIT compiled.
I agree with @dholmes-ora that we should not fix it in Deoptimizer. I think we should restrict "Optimization of Box nodes in uncommon_trap" [8261137 ](https://bugs.openjdk.java.net/browse/JDK-8261137) only to cases when  class is initialized.

**Comment by dcubed-ojdk** [comment]
@vnkozlov and @Wanghuang-Huawei,
This failure appears to be happening twice in every Tier6 job set.
Is there an estimate on when a fix for this issue will be ready?

**Comment by dcubed-ojdk** [comment]
@vnkozlov and @Wanghuang-Huawei,
This failure has now reached Tier8 and we have 8 failures (so far).
I haven't seen a response to my query on Friday. Please ProblemList
this test if the fix is not going to be integrated soon.

**Comment by Wanghuang-Huawei** [comment]
> @vnkozlov and @Wanghuang-Huawei,
> This failure has now reached Tier8 and we have 8 failures (so far).
> I haven't seen a response to my query on Friday. Please ProblemList
> this test if the fix is not going to be integrated soon.

I am very sorry about that.I have fixed this  issue in a new way and  test it on my  x86 and  aarch64 machine. I think I will push a new  commit  tomorrow.

**Comment by Wanghuang-Huawei** [comment]
@dholmes-ora @vnkozlov @dcubed-ojdk @coleenp  I fixed the bug according to @vnkozlov's suggestion. 
I added method `is_box_cache_valid` to judge if a box cache class is ready.

**Review by dholmes-ora — APPROVED** [review]
This approach looks much better to me, but I'll leave it to @vnkozlov to confirm the details.

Thanks,
David

File: src/hotspot/share/opto/callGenerator.cpp [hunk 2]
```diff
@@ -28,6 +28,7 @@
 #include "ci/ciObjArray.hpp"
 #include "ci/ciMemberName.hpp"
 #include "ci/ciMethodHandle.hpp"
+#include "ci/ciUtilities.inline.hpp"
```
Comment (dholmes-ora) [inline src/hotspot/share/opto/callGenerator.cpp:31]: You are also missing the #include of systemDictionary.hpp

Comment (Wanghuang-Huawei) [reply src/hotspot/share/opto/callGenerator.cpp:31]: Thank you for your review. I have fixed this problem.

File: src/hotspot/share/opto/callGenerator.cpp [hunk 3]
```diff
@@ -565,6 +567,32 @@ static bool has_non_debug_usages(Node* n) {
...
+      case T_LONG:    return is_klass_initialized(java_lang_Long_LongCache::symbol());
+      case T_BOOLEAN:
+      case T_FLOAT:
+      case T_DOUBLE:  return true;
+      default:;
+    }
+  }
+  return false;
+}
+
```
Comment (vnkozlov) [inline src/hotspot/share/opto/callGenerator.cpp:595]: The code is correct but, please, move this new code into ciInstanceKlass (near other `box` queries). The VM queries should be done in CI.
```
static bool is_box_cache_valid(CallNode* call) {
  ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
  return klass->is_box_cache_valid();
}
```

Comment (Wanghuang-Huawei) [reply src/hotspot/share/opto/callGenerator.cpp]: I have cleaned these codes. Thank you for your review.

File: src/hotspot/share/ci/ciInstanceKlass.cpp [hunk 4]
```diff
@@ -282,6 +282,31 @@ bool ciInstanceKlass::is_boxed_value_offset(int offset) const {
          (offset == java_lang_boxing_object::value_offset(bt));
 }
 
+static bool is_klass_initialized(Symbol* klass_name) {
+  VM_ENTRY_MARK;
+  InstanceKlass* ik = SystemDictionary::find_instance_klass(klass_name, Handle(), Handle());
```
Comment (dholmes-ora) [inline src/hotspot/share/ci/ciInstanceKlass.cpp:287]: You need the #include for systemDictionary.hpp in this file now.

**Review by dholmes-ora — APPROVED** [review]
Looks fine - just a missing include.

Thanks,
David

File: src/hotspot/share/ci/ciInstanceKlass.cpp (hunk 4 above)
Comment (Wanghuang-Huawei) [reply src/hotspot/share/ci/ciInstanceKlass.cpp:288]: Yes. Thank you for your review. I have adjusted headers in the latest patch.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/runtime/deoptimization.cpp (hunk 1 above)
Comment (Wanghuang-Huawei) [reply src/hotspot/share/runtime/deoptimization.cpp]: Yes. I changed my patch. Thank you for your review.

**Review by dcubed-ojdk — APPROVED** [review]

**Comment by dcubed-ojdk** [comment]
I've reviewed and I'm testing the fix right now. If it passes, then I'll sponsor.

**Comment by dcubed-ojdk** [comment]
I reproduced the failure on my MBP13 with 'fastdebug' and 'slow debug' bits
and then retested with the fix in place with 'fastdebug' bits and the failure
no longer reproduced.

/sponsor

**Pull request closed without merging**